### DISCUSSION TITLE
Price Range랑 Currency랑 동시에 표현하는 State Key 방식

### DISCUSSION THREAD

**Discussion by ysy970923** [Helix-Organization/clp-hooks#2] [General]
Concat해서 사용할까요? Currency, Price Range 이렇게.
그리고 Currency는 1byte, price range도 1byte로 표시할까요?

둘 다 1byte면 충분할 것 같습니다.
그래서 key는 아래처럼 하면 어떨까요?

    void add_liquidity_to_range(uint8_t currency_id, int64_t amount, uint8_t price_range_id) {
        // Logic to add liquidity to the pool at the given price range
        // This will involve updating the state of the pool for the price range
        // and adjusting the total liquidity for that range.
    
        // get the previous amount from the hook state
        uint8_t amount_buffer[4];
        
        // key is the concatenation of the currency and price range
        uint8_t key[2];
        key[0] = currency_id;
        key[1] = price_range_id;
    
        // read the previous amount from the state
        if (state(SBUF(amount_buffer), SBUF(key)) != 16)
        {
            rollback(SBUF("Error reading state"), 1);
        }
        amount += UINT64_FROM_BUF(amount_buffer);
    
        UINT64_TO_BUF(amount_buffer, amount);
        if (state_set(SBUF(amount_buffer), SBUF(key)) != 4)
        {
            rollback(SBUF("Error writing state"), 1);
        }
    }

**Comment by ysy970923** [comment]
price range는 1byte 고려해서 256개로 나눌까요?

**Reply by 0xwonj** [reply]
넵 좋습니다!
그런데 currency_id는 정확히 뭔가요? 토큰은 두개니까 두개 있어야 되는건 아닌가요?

**Reply by ysy970923** [reply]
네! 0,1 이렇게 하거나 아니면 저희가 모든 토큰에 전체 풀들 고려해서 토큰 별로 유니크하게 지정해줄 수도 있을 것 같아요!

**Reply by 0xwonj** [reply]
아 이해 했습니다. 어차피 풀에서 사용하는 토큰은 두가지니까 0 아니면 1로 해도 될 것 같아요.
그리고 constants.h 같은 곳에다가 0이랑 1에 대한 실제 정보를 매핑 시켜놓으면 좋을 것 같습니다.
실제 정보는 일단은 트랜잭션 실제로 받았을 때 이게 0인지 1인지 확인할 때 쓰일 것 같네요.

**Reply by ysy970923** [reply]
네 규칙은 간단하게

지금 교환비가 a:b 라면

a×1/128:b, a×2/128:b, ..., a×128:b 까지로 할까요?

**Reply by ysy970923** [reply]
넵 말씀해주신대로 id는 0,1로 하시죠!

**Comment by 0xwonj** [comment]
그러면 price_range_id 는 예를 들어 0-255 값을 가지게 되는거고,
id 0-255에 해당하는 실제 price range는 저희가 일단 적당히 규칙 세워서 나누고 상수로 정의해놓으면 될까요?

**Comment by ysy970923** [comment]
네 규칙은 간단하게

지금 교환비가 a:b 라면

a×1/128:b, a×2/128:b, ..., a×128:b 까지로 할까요?

2023년 8월 20일 (일) 오전 12:33, wonj ***@***.***>님이 작성:

**Reply by 0xwonj** [reply]
규칙이 정확히 어떻게 되는지 설명해주실 수 있을까요?

**Reply by ysy970923** [reply]
아 아니다 혹시 이렇개 할까요?

교환비를 a라고 했을때 a 의 range가

1.1^(-127)~1.1^(-126), ..., 1.1^(126)~1.1^(127)

로 10프로씩 차이나게 할까요?

**Reply by 0xwonj** [reply]
5프로는 어떠세요? 1.1^127은 18만정도라 너무 의미없이 크고, 1.05는 490정도라 좋은 수치 같아서요
또 간격이 작으면 효율성이 더 높아지지 않을까요?

**Reply by ysy970923** [reply]
넵 좋은것 같습니다.
그러면 5프로이고, 딱 중간 기준은 풀을 만들때 처음 넣은 초기값으로 해서 거기서 5프로씩 위로 127개 밑으로 127개 하면 좋을 것 같습니다!
